common/lru_policy.go

```diff
@@ -0,0 +1,113 @@
+package common
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+)
+
+type CacheBlock struct {
+	sync.RWMutex
+	StartIndex int64
+	EndIndex   int64
+	Data       []byte
+	Last       bool //last block in the file?
+}
+
+//KeyPair: the list node containing both block key and cache block values
+type KeyPair struct {
+	key   int64
+	value *CacheBlock
+}
+
+//LRUCache definition for Least Recently Used Cache implementation
+type LRUCache struct {
+	sync.RWMutex
+	Capacity int64
+	List     *list.List              //DoublyLinkedList: node1->node2.... node:=KeyPair
+	Elements map[int64]*list.Element //blockKey:KeyPair
+	Occupied int64
+}
+
+//NewLRUCache: creates a new LRUCache object with the defined capacity
+func NewLRUCache(capacity int64) *LRUCache {
+	return &LRUCache{
+		Capacity: capacity,
+		List:     new(list.List),
+		Elements: make(map[int64]*list.Element),
+	}
+}
+
+//Get: returns the cache value stored for the key, cache hits the handle and moves the list pointer to front of the list
+func (cache *LRUCache) Get(bk int64) (*CacheBlock, bool) {
+	found := false
```

Comment:
Here we are assuming that we have a start index of the buffer and using that as key we are searching. This might be true in case of read stream but when write stream comes into play this logic will be very specific to read stream alone. So better to rename this method to GetBlockByStartIndex or something like this so that later we can add another method GetBlockByOffset() which write stream can use.

Reply:
why would it be specific? the way I see this is, by the time we get the call here with the startIndex we have already found exactly which block offset from the block list we're looking for, here we are simply seeing if that block is cached or not. But even in write stream we will use start indices.

Reply:
as block sizes are not fixed in case of write streaming, how do you plan to calculate start index from offset?

Reply:
it won't be done on the policy side. It would be done in read_write.go where we will call the binary search on block offset etc when `Get()` in lru policy returns block not found. Then we simply cache the missing block with its block offset list offset/startIndex.